File: FFCoreData/FFCoreData/UITableViewController+FFCoreData.m
//
//  UITableViewController+FFCoreData.m
//  FFCoreData
//
//  Created by Florian Friedrich on 26.09.14.
//  Copyright (c) 2014 Florian Friedrich. All rights reserved.
//

#import "UITableViewController+FFCoreData.h"
#import "FFCoreData.h"
#import <objc/runtime.h>

@implementation UITableViewController (FFCoreData)

- (void)setupWithDelegate:(id<FFCDFetchedResultsControllerDelegate, FFCDTableViewDataSourceDelegate>)delegate {
    return [self setupWithFetchedResultsControllerDelegate:delegate tableViewDataSourceDelegate:delegate];
}

- (void)setupWithFetchedResultsControllerDelegate:(id<FFCDFetchedResultsControllerDelegate>)frcd
                      tableViewDataSourceDelegate:(id<FFCDTableViewDataSourceDelegate>)dataSourceDelegate {
    self.tableView.delegate = self;
    self.fetchedResultsControllerDelegate = [[FFCDTableViewFetchedResultsControllerDelegate alloc]
                                             initWithFetchedResultsController:self.fetchedResultsController
                                             delegate:frcd
                                             tableView:self.tableView];
    self.dataSoure = [[FFCDTableViewDataSource alloc] initWithFetchedResultsController:self.fetchedResultsController
                                                                              delegate:dataSourceDelegate
                                                                             tableView:self.tableView];
}

#pragma mark - Properties
- (NSManagedObjectContext *)managedObjectContext {
    return objc_getAssociatedObject(self, FFCDPropertyKey());
}

- (void)setManagedObjectContext:(NSManagedObjectContext *)managedObjectContext {
    return objc_setAssociatedObject(self, FFCDPropertyKey(), managedObjectContext, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (NSFetchedResultsController *)fetchedResultsController {
    return objc_getAssociatedObject(self, FFCDPropertyKey());
}

- (void)setFetchedResultsController:(NSFetchedResultsController *)fetchedResultsController {
    return objc_setAssociatedObject(self, FFCDPropertyKey(), fetchedResultsController, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (FFCDTableViewFetchedResultsControllerDelegate *)fetchedResultsControllerDelegate {
    return objc_getAssociatedObject(self, FFCDPropertyKey());
}

- (void)setFetchedResultsControllerDelegate:(FFCDTableViewFetchedResultsControllerDelegate *)fetchedResultsControllerDelegate {
    return objc_setAssociatedObject(self, FFCDPropertyKey(), fetchedResultsControllerDelegate, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (FFCDTableViewDataSource *)dataSoure {
    return objc_getAssociatedObject(self, FFCDPropertyKey());
}

- (void)setDataSoure:(FFCDTableViewDataSource *)dataSoure {
    return objc_setAssociatedObject(self, FFCDPropertyKey(), dataSoure, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

@end

File: FFCoreData/FFCoreData/UITableViewController+FFCoreData.swift
//
//  UITableViewController+FFCoreData.swift
//  FFCoreData
//
//  Created by Florian Friedrich on 09/06/15.
//  Copyright © 2015 Florian Friedrich. All rights reserved.
//

#if os(iOS)
import FFCoreData
import UIKit
    
public extension UITableViewController {
    private struct PropertyKeys {
        static var ManagedObjectContext = "UITableViewController+FFCoreData.ManagedObjectContext"
        static var FetchedResultsController = "UITableViewController+FFCoreData.FetchedResultsController"
        static var DataSource = "UITableViewController+FFCoreData.DataSource"
        static var FetchedResultsControllerDelegate = "UITableViewController+FFCoreData.FetchedResultsControllerDelegate"
    }
    /**
    *  The managed object context of the UITableViewController (used for the NSFetchedResultsController).
    */
    public var managedObjectContext: NSManagedObjectContext {
        get {
            if let moc = objc_getAssociatedObject(self, &PropertyKeys.ManagedObjectContext) as? NSManagedObjectContext
            {
                return moc
            } else {
                self.managedObjectContext = CoreDataStack.MainContext
                return self.managedObjectContext
            }
        }
        set {
            objc_setAssociatedObject(self, &PropertyKeys.ManagedObjectContext, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }
    
    /**
    *  The fetched results controller for the tableview.
    */
    public var fetchedResultsController: NSFetchedResultsController? {
        get {
            return objc_getAssociatedObject(self, &PropertyKeys.FetchedResultsController) as? NSFetchedResultsController
        }
        set {
            objc_setAssociatedObject(self, &PropertyKeys.FetchedResultsController, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }
    
    /**
    *  The delegate of the fetched results controller.
    */
    public var fetchedResultsControllDelegate: TableViewFetchedResultsControllerDelegate? {
        get {
            return objc_getAssociatedObject(self, &PropertyKeys.FetchedResultsControllerDelegate) as? TableViewFetchedResultsControllerDelegate
        }
        set {
            objc_setAssociatedObject(self, &PropertyKeys.FetchedResultsControllerDelegate, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }

    /**
    *  The data source for the tableview.
    */
    public var dataSource: TableViewDataSource? {
        get {
            return objc_getAssociatedObject(self, &PropertyKeys.DataSource) as? TableViewDataSource
        }
        set {
            objc_setAssociatedObject(self, &PropertyKeys.DataSource, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }
    
    /**
    *  Sets up the fetchedResultsControllerDelegate and the dataSource with its delegates.
    *  The managedObjectContext and the fetchedResultsController must be set before this method is called!
    *  @param frcDelegate        The delegate for the FFCDTableViewFetchedResultsControllerDelegate.
    *  @param dataSourceDelegate The delegate for the FFCDTableViewDataSource.
    */
    func setupWithFetchedResultsControllerDelegate(frcDelegate: FetchedResultsControllerDelegateDelegate, dataSourceDelegate: TableViewDataSourceDelegate) {
        tableView.delegate = self
        if let frc = fetchedResultsController, tableView = tableView {
            fetchedResultsControllDelegate = TableViewFetchedResultsControllerDelegate(fetchedResultsController: frc, tableView: tableView, delegate: frcDelegate)
            dataSource = TableViewDataSource(tableView: tableView, controller: frc, delegate: dataSourceDelegate)
        }
    }
    
    /**
    *  Sets up the fetchedResultsControllerDelegate and the dataSource with the same delegate.
    *  @param delegate The delegate for both, the fetchedResultsControllerDelegate and the dataSource.
    */
    func setupWithDelegate(delegate: protocol<FetchedResultsControllerDelegateDelegate, TableViewDataSourceDelegate>) {
        setupWithFetchedResultsControllerDelegate(delegate, dataSourceDelegate: delegate)
    }
}
#endif
